File: src/swe_forge/execution/container.py
"""Container lifecycle management for Docker execution.

This module provides high-level abstractions for managing container
lifecycle including creation, execution, and cleanup with guaranteed
resource cleanup on exceptions.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timezone
from enum import Enum
from logging import getLogger
from typing import TYPE_CHECKING, Any, Self

if TYPE_CHECKING:
    from collections.abc import AsyncGenerator

from swe_forge.execution.docker_client import (
    ContainerConfig,
    ContainerStatus,
    DockerClient,
    DockerError,
    ExecResult,
)

logger = getLogger(__name__)


class ManagedContainerStatus(Enum):
    """Status of a managed container."""

    PENDING = "pending"
    CREATING = "creating"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    TIMEOUT = "timeout"

    def __str__(self) -> str:
        return self.value


@dataclass
class VolumeMount:
    """Configuration for mounting a host path into a container.

    Attributes:
        host_path: Path on the host system to mount.
        container_path: Path inside the container where the mount appears.
        read_only: If True, mount is read-only.
    """

    host_path: str
    container_path: str
    read_only: bool = False

    def to_docker_bind(self) -> str:
        """Convert to Docker bind mount format.

        Returns:
            Docker bind mount string like "/host:/container:ro" or "/host:/container:rw"
        """
        mode = "ro" if self.read_only else "rw"
        return f"{self.host_path}:{self.container_path}:{mode}"


@dataclass
class ContainerSpec:
    """Specification for creating and managing a container.

    This is a higher-level configuration than ContainerConfig, providing
    more intuitive interfaces for common operations like volume mounts
    and environment variables.

    Attributes:
        name: Container name.
        image: Docker image to use.
        command: Command to run in the container.
        volumes: List of volume mounts.
        env: Dictionary of environment variables.
        working_dir: Working directory inside the container.
        user: User to run as (e.g., "1000:1000").
        network_mode: Network mode for the container.
        memory_mb: Memory limit in megabytes.
        cpu_limit: CPU limit as fraction (1.0 = 1 CPU).
        pids_limit: Maximum number of processes.
        stop_timeout: Seconds to wait for graceful shutdown before SIGKILL.
    """

    name: str
    image: str
    command: list[str] | None = None
    volumes: list[VolumeMount] = field(default_factory=list)
    env: dict[str, str] = field(default_factory=dict)
    working_dir: str | None = None
    user: str | None = None
    network_mode: str = "bridge"
    memory_mb: int = 512
    cpu_limit: float = 1.0
    pids_limit: int = 100
    stop_timeout: int = 10

    def to_container_config(self) -> ContainerConfig:
        """Convert to ContainerConfig for DockerClient.

        Returns:
            ContainerConfig instance ready for DockerClient.create_container.
        """
        # Convert volume mounts to Docker bind format
        volume_binds = [v.to_docker_bind() for v in self.volumes]

        # Convert env dict to list of "KEY=value" strings
        env_list = [f"{k}={v}" for k, v in self.env.items()]

        return ContainerConfig(
            name=self.name,
            image=self.image,
            cmd=self.command,
            env=env_list,
            working_dir=self.working_dir,
            volumes=volume_binds,
            user=self.user,
            network_mode=self.network_mode,
            memory_mb=self.memory_mb,
            cpu_limit=self.cpu_limit,
            pids_limit=self.pids_limit,
        )


@dataclass
class ManagedContainer:
    """A managed Docker container with lifecycle tracking.

    This class tracks the state of a container and provides
    metadata about its creation and execution.
    """

    id: str
    spec: ContainerSpec
    status: ManagedContainerStatus = ManagedContainerStatus.PENDING
    created_at: datetime = field(default_factory=lambda: datetime.now(timezone.utc))
    exit_code: int | None = None
    error_message: str | None = None

    @property
    def is_terminal(self) -> bool:
        """Check if the container is in a terminal state."""
        return self.status in (
            ManagedContainerStatus.COMPLETED,
            ManagedContainerStatus.FAILED,
            ManagedContainerStatus.TIMEOUT,
        )

    @property
    def is_running(self) -> bool:
        """Check if the container is currently running."""
        return self.status == ManagedContainerStatus.RUNNING


class ContainerManager:
    """Async context manager for Docker container lifecycle.

    Provides guaranteed cleanup on both normal exit and exceptions,
    implementing the RAII pattern for container management.

    Usage:
        spec = ContainerSpec(
            name="test-container",
            image="python:latest"  # Agent determines,
            volumes=[VolumeMount("/host/path", "/container/path")],
            env={"FOO": "bar"},
        )

        async with ContainerManager(docker_client, spec) as container:
            # Container is running
            result = await container.exec(["ls", "-la"])
            # Automatic cleanup on exit
    """

    def __init__(
        self,
        client: DockerClient,
        spec: ContainerSpec,
        *,
        auto_start: bool = True,
        auto_remove: bool = True,
    ) -> None:
        """Initialize the container manager.

        Args:
            client: DockerClient instance for Docker operations.
            spec: Container specification.
            auto_start: If True, start the container after creation.
            auto_remove: If True, remove container on cleanup.
        """
        self._client = client
        self._spec = spec
        self._auto_start = auto_start
        self._auto_remove = auto_remove
        self._container: ManagedContainer | None = None
        self._started = False

    @classmethod
    def from_existing(
        cls,
        client: DockerClient,
        container_id: str,
        spec: ContainerSpec,
        *,
        auto_remove: bool = True,
    ) -> Self:
        """Create a manager for an existing container.

        This is useful for reconnecting to containers created elsewhere.

        Args:
            client: DockerClient instance for Docker operations.
            container_id: ID of an existing container.
            spec: Container specification (for reference).
            auto_remove: If True, remove container on cleanup.

        Returns:
            ContainerManager instance attached to existing container.
        """
        manager = cls(client, spec, auto_start=False, auto_remove=auto_remove)
        manager._container = ManagedContainer(
            id=container_id,
            spec=spec,
            status=ManagedContainerStatus.PENDING,
        )
        return manager

    async def __aenter__(self) -> Self:
        """Create and optionally start the container.

        Returns:
            Self (ContainerManager) for method chaining.

        Raises:
            DockerError: If container creation or start fails.
        """
        try:
            await self._client.ensure_image(self._spec.image)

            config = self._spec.to_container_config()
            container_id = await self._client.create_container(config)

            self._container = ManagedContainer(
                id=container_id,
                spec=self._spec,
                status=ManagedContainerStatus.PENDING,
            )

            if self._auto_start:
                await self.start()

            return self

        except Exception:
            await self._cleanup(force=True)
            raise

    async def __aexit__(
        self,
        exc_type: type[BaseException] | None,
        exc_val: BaseException | None,
        exc_tb: Any,
    ) -> None:
        """Cleanup the container on exit.

        Guarantees container removal even on exceptions.
        Logs warnings if cleanup fails but doesn't mask original exception.
        """
        await self._cleanup(force=exc_type is not None)

    async def _cleanup(self, force: bool = False) -> None:
        if self._container is None:
            return

        container_id = self._container.id

        try:
            if self._started:
                try:
                    await self._client.stop_container(
                        container_id, timeout=self._spec.stop_timeout
                    )
                    self._container.status = ManagedContainerStatus.COMPLETED
                except Exception:
                    force = True

            if self._auto_remove:
                try:
                    await self._client.remove_container(container_id, force=force)
                except Exception:
                    pass

            self._container = None
            self._started = False

        except Exception as e:
            logger.error(f"Container cleanup failed for {container_id}: {e}")

    async def start(self) -> None:
        """Start the container.

        Raises:
            DockerError: If container is not in PENDING state or start fails.
        """
        if self._container is None:
            raise DockerError("Container not created")

        if self._container.status != ManagedContainerStatus.PENDING:
            raise DockerError(
                f"Cannot start container in {self._container.status} state",
                container_id=self._container.id,
            )

        self._container.status = ManagedContainerStatus.CREATING

        try:
            await self._client.start_container(self._container.id)
            self._container.status = ManagedContainerStatus.RUNNING
            self._started = True
        except DockerError as e:
            self._container.status = ManagedContainerStatus.FAILED
            self._container.error_message = str(e)
            raise

    async def stop(self) -> None:
        """Stop the container gracefully.

        Uses the stop_timeout from the spec.
        """
        if self._container is None or not self._started:
            return

        if self._container.status != ManagedContainerStatus.RUNNING:
            return

        try:
            await self._client.stop_container(
                self._container.id, timeout=self._spec.stop_timeout
            )
            self._container.status = ManagedContainerStatus.COMPLETED
            self._started = False
        except DockerError as e:
            self._container.status = ManagedContainerStatus.FAILED
            self._container.error_message = str(e)
            raise

    async def exec(
        self,
        cmd: list[str],
        *,
        timeout: float | None = None,
    ) -> ExecResult:
        """Execute a command in the container.

        Args:
            cmd: Command to execute as list of strings.
            timeout: Optional timeout in seconds.

        Returns:
            ExecResult with exit_code, stdout, and stderr.

        Raises:
            DockerError: If container is not running or exec fails.
        """
        if self._container is None:
            raise DockerError("Container not created")

        if (
            not self._started
            or self._container.status != ManagedContainerStatus.RUNNING
        ):
            raise DockerError(
                f"Cannot exec in container with {self._container.status} state",
                container_id=self._container.id,
            )

        result = await self._client.exec(self._container.id, cmd, timeout=timeout)
        return result

    async def get_logs(
        self,
        *,
        stdout: bool = True,
        stderr: bool = True,
        tail: int | None = None,
    ) -> str:
        """Get logs from the container.

        Args:
            stdout: Include stdout.
            stderr: Include stderr.
            tail: Number of lines to return from end.

        Returns:
            Combined logs as string.
        """
        if self._container is None:
            raise DockerError("Container not created")

        return await self._client.get_logs(
            self._container.id,
            stdout=stdout,
            stderr=stderr,
            tail=tail,
        )

    async def stream_logs(
        self,
        *,
        stdout: bool = True,
        stderr: bool = True,
    ) -> AsyncGenerator[str, None]:
        """Stream logs from the container.

        Args:
            stdout: Include stdout.
            stderr: Include stderr.

        Yields:
            Log chunks as strings.
        """
        if self._container is None:
            raise DockerError("Container not created")

        async for chunk in self._client.stream_logs(
            self._container.id,
            stdout=stdout,
            stderr=stderr,
        ):
            yield chunk

    async def wait(self, *, timeout: float | None = None) -> int:
        """Wait for the container to finish.

        Args:
            timeout: Optional timeout in seconds.

        Returns:
            Exit code of the container.

        Raises:
            DockerError: If timeout is reached.
        """
        if self._container is None:
            raise DockerError("Container not created")

        exit_code = await self._client.wait_container(
            self._container.id, timeout=timeout
        )
        self._container.exit_code = exit_code
        self._started = False

        if exit_code == 0:
            self._container.status = ManagedContainerStatus.COMPLETED
        else:
            self._container.status = ManagedContainerStatus.FAILED
            self._container.error_message = f"Exited with code {exit_code}"

        return exit_code

    async def sync_status(self) -> ManagedContainerStatus:
        """Sync local status with actual Docker container status.

        Returns:
            Updated container status.
        """
        if self._container is None:
            raise DockerError("Container not created")

        try:
            status = await self._client.get_container_status(self._container.id)

            status_map = {
                ContainerStatus.CREATED: ManagedContainerStatus.PENDING,
                ContainerStatus.RUNNING: ManagedContainerStatus.RUNNING,
                ContainerStatus.PAUSED: ManagedContainerStatus.RUNNING,
                ContainerStatus.RESTARTING: ManagedContainerStatus.RUNNING,
                ContainerStatus.EXITED: ManagedContainerStatus.COMPLETED,
                ContainerStatus.REMOVING: ManagedContainerStatus.COMPLETED,
                ContainerStatus.DEAD: ManagedContainerStatus.FAILED,
                ContainerStatus.UNKNOWN: ManagedContainerStatus.FAILED,
            }

            self._container.status = status_map.get(
                status, ManagedContainerStatus.FAILED
            )

            return self._container.status

        except DockerError:
            self._container.status = ManagedContainerStatus.FAILED
            return self._container.status

    @property
    def container(self) -> ManagedContainer | None:
        """Get the managed container, or None if not created."""
        return self._container

    @property
    def container_id(self) -> str | None:
        """Get the container ID, or None if not created."""
        return self._container.id if self._container else None

File: src/execution/container.rs
//! Container lifecycle management for Docker execution.
//!
//! This module provides a high-level abstraction for managing container
//! lifecycle including creation, execution, and cleanup.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use crate::error::DockerError;
use crate::execution::docker_client::{ContainerConfig, ContainerStatusInfo, DockerClient};

// Re-export ExecResult from docker_client for convenience
pub use crate::execution::docker_client::ExecResult;

/// Status of a managed container.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum ContainerStatus {
    /// Container is pending creation.
    Pending,
    /// Container is being created.
    Creating,
    /// Container is running.
    Running,
    /// Container completed successfully.
    Completed,
    /// Container failed with an error message.
    Failed(String),
    /// Container exceeded its timeout.
    Timeout,
}

impl std::fmt::Display for ContainerStatus {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ContainerStatus::Pending => write!(f, "pending"),
            ContainerStatus::Creating => write!(f, "creating"),
            ContainerStatus::Running => write!(f, "running"),
            ContainerStatus::Completed => write!(f, "completed"),
            ContainerStatus::Failed(msg) => write!(f, "failed: {}", msg),
            ContainerStatus::Timeout => write!(f, "timeout"),
        }
    }
}

/// A managed Docker container with lifecycle tracking.
#[derive(Debug)]
pub struct Container {
    /// Docker container ID.
    id: String,
    /// Current status of the container.
    status: ContainerStatus,
    /// Configuration used to create the container.
    config: ContainerConfig,
    /// Timestamp when the container was created.
    created_at: DateTime<Utc>,
}

impl Container {
    /// Creates a new container with the given configuration.
    ///
    /// This creates the container in Docker but does not start it.
    /// The container starts in `Creating` status and transitions to
    /// `Pending` when successfully created.
    ///
    /// # Arguments
    ///
    /// * `client` - Docker client for API operations
    /// * `config` - Container configuration
    ///
    /// # Errors
    ///
    /// Returns `DockerError` if container creation fails.
    ///
    /// # Example
    ///
    /// ```ignore
    /// use swe_forge::execution::{DockerClient, Container, ContainerConfig};
    ///
    /// let client = DockerClient::new()?;
    /// let config = ContainerConfig::new("task-123", "python:3.11-slim");
    /// let container = Container::new(&client, config).await?;
    /// ```
    pub async fn new(client: &DockerClient, config: ContainerConfig) -> Result<Self, DockerError> {
        let created_at = Utc::now();

        // Ensure image exists
        if !client.image_exists(&config.image).await {
            client.pull_image(&config.image).await?;
        }

        let id = client.create_container(config.clone()).await?;

        Ok(Self {
            id,
            status: ContainerStatus::Pending,
            config,
            created_at,
        })
    }

    /// Creates a container struct from an existing Docker container ID.
    ///
    /// This is useful for reconnecting to containers that were created
    /// in a previous session.
    ///
    /// # Arguments
    ///
    /// * `client` - Docker client for API operations
    /// * `id` - Existing container ID
    /// * `config` - Configuration (for reference, not applied)
    pub async fn from_existing(
        client: &DockerClient,
        id: impl Into<String>,
        config: ContainerConfig,
    ) -> Result<Self, DockerError> {
        let id = id.into();

        // Verify container exists and get its status
        let status_info = client.container_status(&id).await?;
        let status = match status_info {
            ContainerStatusInfo::Created => ContainerStatus::Pending,
            ContainerStatusInfo::Running => ContainerStatus::Running,
            ContainerStatusInfo::Exited { exit_code } => {
                if exit_code == 0 {
                    ContainerStatus::Completed
                } else {
                    ContainerStatus::Failed(format!("Exited with code {}", exit_code))
                }
            }
            ContainerStatusInfo::Dead => ContainerStatus::Failed("Container is dead".to_string()),
            _ => ContainerStatus::Pending,
        };

        Ok(Self {
            id,
            status,
            config,
            created_at: Utc::now(), // We don't know the original creation time
        })
    }

    /// Starts the container.
    ///
    /// Transitions the container from `Pending` to `Running` status.
    ///
    /// # Errors
    ///
    /// Returns `DockerError` if:
    /// - Container is not in `Pending` status
    /// - Docker API call fails
    pub async fn start(&mut self, client: &DockerClient) -> Result<(), DockerError> {
        if self.status != ContainerStatus::Pending {
            return Err(DockerError::RunFailed(format!(
                "Cannot start container in {} state",
                self.status
            )));
        }

        self.status = ContainerStatus::Creating;

        match client.start_container(&self.id).await {
            Ok(()) => {
                self.status = ContainerStatus::Running;
                Ok(())
            }
            Err(e) => {
                self.status = ContainerStatus::Failed(format!("Start failed: {}", e));
                Err(e)
            }
        }
    }

    /// Executes a command inside the running container.
    ///
    /// # Arguments
    ///
    /// * `client` - Docker client for API operations
    /// * `cmd` - Command to execute as array of strings
    ///
    /// # Returns
    ///
    /// `ExecResult` containing exit code, stdout, and stderr.
    ///
    /// # Errors
    ///
    /// Returns `DockerError` if:
    /// - Container is not running
    /// - Command execution fails
    ///
    /// # Example
    ///
    /// ```ignore
    /// let result = container.exec(&client, &["python", "-c", "print('hello')"]).await?;
    /// println!("Output: {}", result.stdout);
    /// ```
    pub async fn exec(
        &self,
        client: &DockerClient,
        cmd: &[&str],
    ) -> Result<ExecResult, DockerError> {
        if self.status != ContainerStatus::Running {
            return Err(DockerError::RunFailed(format!(
                "Cannot exec in container with {} state",
                self.status
            )));
        }

        client.exec_command(&self.id, cmd).await
    }

    /// Stops the container if running.
    ///
    /// This sends SIGTERM to the container and waits for graceful shutdown.
    pub async fn stop(&mut self, client: &DockerClient) -> Result<(), DockerError> {
        if self.status != ContainerStatus::Running {
            return Ok(()); // Already stopped
        }

        client.stop_container(&self.id).await?;

        // Check final status
        let status_info = client.container_status(&self.id).await?;
        self.status = match status_info {
            ContainerStatusInfo::Exited { exit_code } => {
                if exit_code == 0 {
                    ContainerStatus::Completed
                } else {
                    ContainerStatus::Failed(format!("Exited with code {}", exit_code))
                }
            }
            _ => ContainerStatus::Completed,
        };

        Ok(())
    }

    /// Cleans up the container by stopping and removing it.
    ///
    /// This is the final step in the container lifecycle and should
    /// be called when the container is no longer needed.
    ///
    /// # Arguments
    ///
    /// * `client` - Docker client for API operations
    ///
    /// # Errors
    ///
    /// Returns `DockerError` if cleanup fails, though it attempts
    /// to force removal if graceful cleanup fails.
    pub async fn cleanup(&mut self, client: &DockerClient) -> Result<(), DockerError> {
        // Try graceful stop first if running
        if self.status == ContainerStatus::Running {
            if let Err(e) = client.stop_container(&self.id).await {
                // Log error but continue with force removal
                tracing::warn!("Failed to stop container gracefully: {}", e);
            }
        }

        // Force remove the container
        client.remove_container(&self.id, true).await?;

        // Update status based on previous state
        match &self.status {
            ContainerStatus::Running | ContainerStatus::Pending | ContainerStatus::Creating => {
                self.status = ContainerStatus::Completed;
            }
            // Keep Failed/Timeout/Completed status
            _ => {}
        }

        Ok(())
    }

    /// Marks the container as timed out and cleans it up.
    pub async fn mark_timeout(&mut self, client: &DockerClient) -> Result<(), DockerError> {
        self.status = ContainerStatus::Timeout;
        self.cleanup(client).await
    }

    /// Marks the container as failed with a message.
    pub fn mark_failed(&mut self, message: impl Into<String>) {
        self.status = ContainerStatus::Failed(message.into());
    }

    /// Gets the logs from the container.
    pub async fn logs(&self, client: &DockerClient) -> Result<String, DockerError> {
        client.get_logs(&self.id).await
    }

    /// Waits for the container to finish executing.
    ///
    /// # Returns
    ///
    /// The exit code of the container.
    pub async fn wait(&mut self, client: &DockerClient) -> Result<i64, DockerError> {
        let exit_code = client.wait_container(&self.id).await?;

        self.status = if exit_code == 0 {
            ContainerStatus::Completed
        } else {
            ContainerStatus::Failed(format!("Exited with code {}", exit_code))
        };

        Ok(exit_code)
    }

    /// Returns the container ID.
    pub fn id(&self) -> &str {
        &self.id
    }

    /// Returns the current status.
    pub fn status(&self) -> &ContainerStatus {
        &self.status
    }

    /// Returns the container configuration.
    pub fn config(&self) -> &ContainerConfig {
        &self.config
    }

    /// Returns when the container was created.
    pub fn created_at(&self) -> DateTime<Utc> {
        self.created_at
    }

    /// Returns the configured timeout in seconds.
    pub fn timeout_seconds(&self) -> u64 {
        self.config.limits.timeout_seconds
    }

    /// Checks if the container is in a terminal state (completed, failed, timeout).
    pub fn is_terminal(&self) -> bool {
        matches!(
            self.status,
            ContainerStatus::Completed | ContainerStatus::Failed(_) | ContainerStatus::Timeout
        )
    }

    /// Checks if the container is running.
    pub fn is_running(&self) -> bool {
        self.status == ContainerStatus::Running
    }

    /// Syncs the local status with the actual Docker container status.
    pub async fn sync_status(&mut self, client: &DockerClient) -> Result<(), DockerError> {
        let status_info = client.container_status(&self.id).await?;

        self.status = match status_info {
            ContainerStatusInfo::Created => ContainerStatus::Pending,
            ContainerStatusInfo::Running => ContainerStatus::Running,
            ContainerStatusInfo::Paused => ContainerStatus::Running, // Treat paused as running
            ContainerStatusInfo::Restarting => ContainerStatus::Running,
            ContainerStatusInfo::Exited { exit_code } => {
                if exit_code == 0 {
                    ContainerStatus::Completed
                } else {
                    ContainerStatus::Failed(format!("Exited with code {}", exit_code))
                }
            }
            ContainerStatusInfo::Removing => ContainerStatus::Completed,
            ContainerStatusInfo::Dead => ContainerStatus::Failed("Container is dead".to_string()),
            ContainerStatusInfo::Unknown(s) => {
                ContainerStatus::Failed(format!("Unknown status: {}", s))
            }
        };

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::execution::ExecutionLimits;

    #[test]
    fn test_container_status_display() {
        assert_eq!(format!("{}", ContainerStatus::Pending), "pending");
        assert_eq!(format!("{}", ContainerStatus::Creating), "creating");
        assert_eq!(format!("{}", ContainerStatus::Running), "running");
        assert_eq!(format!("{}", ContainerStatus::Completed), "completed");
        assert_eq!(format!("{}", ContainerStatus::Timeout), "timeout");
        assert_eq!(
            format!("{}", ContainerStatus::Failed("error".to_string())),
            "failed: error"
        );
    }

    #[test]
    fn test_container_status_equality() {
        assert_eq!(ContainerStatus::Running, ContainerStatus::Running);
        assert_ne!(ContainerStatus::Running, ContainerStatus::Completed);
        assert_eq!(
            ContainerStatus::Failed("a".to_string()),
            ContainerStatus::Failed("a".to_string())
        );
        assert_ne!(
            ContainerStatus::Failed("a".to_string()),
            ContainerStatus::Failed("b".to_string())
        );
    }

    #[test]
    fn test_is_terminal() {
        // Terminal states
        assert!(ContainerStatus::Completed.to_string().contains("completed"));
        assert!(ContainerStatus::Timeout.to_string().contains("timeout"));

        // Non-terminal states
        assert!(!ContainerStatus::Pending.to_string().contains("terminal"));
        assert!(!ContainerStatus::Running.to_string().contains("terminal"));
    }

    #[test]
    fn test_container_config_limits() {
        let limits = ExecutionLimits::new(2048, 2.0, 10, 200, 1800);
        let config = ContainerConfig::new("test", "python:3.11").with_limits(limits);

        assert_eq!(config.limits.memory_mb, 2048);
        assert_eq!(config.limits.timeout_seconds, 1800);
    }
}
